File: modules/app/agent-repl/daemon/internal/sessioncontroller/conversationpage_test.go
package sessioncontroller

import (
	"context"
	"errors"
	"fmt"
	"testing"

	corev1 "agentrepl/proto/agentshim/core/v1"
	datav1 "agentrepl/proto/agentshim/data/v1"
	frontendv1 "agentrepl/proto/agentshim/frontend/v1"

	"claude-repld/internal/errclass"
	"claude-repld/internal/ssm"

	"google.golang.org/protobuf/types/known/anypb"
)

// THE COLD OPEN IS NOT A FULL REPLAY ANY MORE.
//
// Every case here is about one of the three promises a page makes that the
// from_seq=0 replay it replaces did not have to: it is BOUNDED (a limit that
// counts renderable things, and a ceiling a client cannot talk its way past),
// it is CONTINUABLE (a cursor that reads the page before, and a start arm that
// says there is no such page), and it SPLICES GAP-FREE onto the live stream
// (live_join_seq, which is the seq the client rejoins at rather than whatever
// seq the first delta happens to carry).

// pageHarness is a durableHarness whose workspace has an authoritative
// WorkspaceState, which is the identity a page's fence is measured against.
type pageHarness struct {
	*durableHarness
}

// newPageHarness arranges an UNWIRED workspace — the state every workspace is
// in after a daemon bounce, and the one a cold webview actually mounts
// against — holding the given store events.
func newPageHarness(t *testing.T, events []*corev1.Event) *pageHarness {
	t.Helper()
	h := &pageHarness{durableHarness: newDurableHarness(t, &durableHistorySpy{events: events})}
	// The Fence is stamped exactly as the mint stamps it (ssm.compositeWorkspaceState),
	// because it is the token the reader compares against and the one a client
	// was actually shown. A fixture that carried the two identities but no fence
	// would be describing a WorkspaceState the daemon never publishes.
	h.applier.current = map[string]*frontendv1.WorkspaceState{
		"ws": {Workspace: "ws", SessionId: "s1", ControllerGenerationId: "g1", Fence: ssm.Fence("s1", "g1")},
	}
	if len(events) > 0 {
		h.seq.SetLastSeq("s1", events[len(events)-1].GetSeq())
	}
	return h
}

// page asks for one page under the live identity.
func (h *pageHarness) page(t *testing.T, anchor PageAnchor) *frontendv1.ConversationPage {
	t.Helper()
	got, err := h.m.ConversationPage(context.Background(), "ws", ssm.Fence("s1", "g1"), anchor)
	if err != nil {
		t.Fatalf("ConversationPage(%+v): %v", anchor, err)
	}
	return got
}

// pageAssistantEvent is a vendor assistant record carrying `blocks`, at seq.
//
// The blocks matter to exactly one question this file asks: how many TOP-LEVEL
// items an event contributes. A record with a text block and a tool_use block
// is ONE feed item carrying its tool call inside it, not two.
func pageAssistantEvent(t *testing.T, seq uint64, uuid string, blocks []*datav1.ContentBlock) *corev1.Event {
	t.Helper()
	a, err := anypb.New(&datav1.ClaudeStreamMessage{
		Msg: &datav1.ClaudeStreamMessage_Assistant{Assistant: &datav1.AssistantMessage{
			Uuid:    uuid,
			Message: &datav1.ApiAssistantMessage{Content: blocks},
		}},
	})
	if err != nil {
		t.Fatalf("anypb.New: %v", err)
	}
	return &corev1.Event{SessionId: "vendor-uuid", Seq: seq, ProducedAtMs: int64(seq) * 1_000, Payload: &corev1.Event_Vendor{Vendor: a}}
}

func pageTextBlock(text string) *datav1.ContentBlock {
	return &datav1.ContentBlock{Block: &datav1.ContentBlock_Text{Text: &datav1.TextBlock{Text: text}}}
}

func pageToolUseBlock(id, name string) *datav1.ContentBlock {
	return &datav1.ContentBlock{Block: &datav1.ContentBlock_ToolUse{ToolUse: &datav1.ToolUseBlock{Id: id, Name: name}}}
}

// pageTextEvents is a run of `n` one-item events at seqs 1..n.
func pageTextEvents(t *testing.T, n int) []*corev1.Event {
	t.Helper()
	var out []*corev1.Event
	for i := 1; i <= n; i++ {
		out = append(out, pageAssistantEvent(t, uint64(i), fmt.Sprintf("u%d", i), []*datav1.ContentBlock{pageTextBlock(fmt.Sprintf("m%d", i))}))
	}
	return out
}

// itemUUIDs names what a page actually carried, which is what every ordering
// and boundary assertion below reads.
func pageItemUUIDs(page *frontendv1.ConversationPage) []string {
	var out []string
	for _, it := range page.GetItems() {
		out = append(out, it.GetUuid())
	}
	return out
}

// --- the tail page ----------------------------------------------------------

func TestATailPageServesTheNewestItemsOldestFirst(t *testing.T) {
	// Arrange — twenty items in the store, of which a cold open wants the last
	// three.
	h := newPageHarness(t, pageTextEvents(t, 20))

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert — the newest three, in feed order.
	got := pageItemUUIDs(page)
	if len(got) != 3 || got[0] != "u18" || got[1] != "u19" || got[2] != "u20" {
		t.Fatalf("tail page items = %v, want [u18 u19 u20]", got)
	}
}

func TestATailPageJoinsTheLiveStreamAtItsNewestSeq(t *testing.T) {
	// Arrange — the page's newest item sits at seq 20. A client that stored
	// "whatever seq the first delta carries" instead would skip everything the
	// session produced between the mint and its subscribe.
	h := newPageHarness(t, pageTextEvents(t, 20))

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert — the resync mark is INCLUSIVE, so rejoining at 20 re-delivers
	// item 20 (deduped by uuid) and everything above it.
	if page.GetLiveJoinSeq() != 20 {
		t.Fatalf("live_join_seq = %d, want 20", page.GetLiveJoinSeq())
	}
}

func TestABeforePageCarriesNoLiveJoinSeq(t *testing.T) {
	// Arrange — a load-more page is history; it has no live edge to join at,
	// and a non-zero mark here would move a client's cursor BACKWARDS.
	h := newPageHarness(t, pageTextEvents(t, 20))
	tail := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Act.
	older := h.page(t, PageAnchor{Cursor: tail.GetMore().GetCursor(), Limit: 3})

	// Assert.
	if older.GetLiveJoinSeq() != 0 {
		t.Fatalf("before page live_join_seq = %d, want 0", older.GetLiveJoinSeq())
	}
}

// --- boundary classification ------------------------------------------------

func TestConstituentsRideInsideTheirItemAndDoNotCountTowardTheLimit(t *testing.T) {
	// Arrange — three records, each ONE top-level item, and each carrying a
	// tool call inside it. A page that counted constituents would serve one
	// item for a limit of three.
	var events []*corev1.Event
	for i := 1; i <= 3; i++ {
		events = append(events, pageAssistantEvent(t, uint64(i), fmt.Sprintf("u%d", i), []*datav1.ContentBlock{
			pageTextBlock(fmt.Sprintf("m%d", i)),
			pageToolUseBlock(fmt.Sprintf("tool-%d", i), "Read"),
		}))
	}
	h := newPageHarness(t, events)

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert — three renderable things, each still carrying its tool call.
	got := pageItemUUIDs(page)
	if len(got) != 3 {
		t.Fatalf("page items = %v, want 3 top-level items with their tool calls inside them", got)
	}
	blocks := page.GetItems()[0].GetAgent().GetResponse().GetBody().GetContent()
	if len(blocks) != 2 {
		t.Fatalf("first item carried %d content block(s), want the text and the tool_use it was emitted with", len(blocks))
	}
}

// --- continuation -----------------------------------------------------------

func TestAPageWithOlderHistoryHandsBackAContinuationCursor(t *testing.T) {
	// Arrange.
	h := newPageHarness(t, pageTextEvents(t, 20))

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert.
	if page.GetMore().GetCursor() == "" {
		t.Fatalf("page continuation = %T, want a `more` arm with a cursor", page.GetContinuation())
	}
}

func TestTheContinuationCursorReadsThePageImmediatelyOlder(t *testing.T) {
	// Arrange — the tail page served u18..u20.
	h := newPageHarness(t, pageTextEvents(t, 20))
	tail := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Act.
	older := h.page(t, PageAnchor{Cursor: tail.GetMore().GetCursor(), Limit: 3})

	// Assert — the three immediately before, with no overlap and no gap.
	got := pageItemUUIDs(older)
	if len(got) != 3 || got[0] != "u15" || got[1] != "u16" || got[2] != "u17" {
		t.Fatalf("second page items = %v, want [u15 u16 u17]", got)
	}
}

func TestAPageThatReachesTheBeginningRetiresLoadMore(t *testing.T) {
	// Arrange — exactly three items exist, and the page asks for three. The
	// walk reaches the floor with nothing left behind, which is the one case
	// that is genuinely the beginning.
	h := newPageHarness(t, pageTextEvents(t, 3))

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert.
	if page.GetStart() == nil {
		t.Fatalf("page continuation = %v, want the `start` arm so the client retires load-more", page.GetContinuation())
	}
}

func TestAnEmptyConversationPagesToTheBeginningWithNoItems(t *testing.T) {
	// Arrange — a workspace whose store holds nothing. Silence must be
	// reported as the empty conversation it is, never as a failure and never
	// as an endless load-more.
	h := newPageHarness(t, nil)

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 10})

	// Assert.
	if len(page.GetItems()) != 0 {
		t.Fatalf("page items = %v, want none", pageItemUUIDs(page))
	}
	if page.GetStart() == nil {
		t.Fatalf("page continuation = %v, want the `start` arm", page.GetContinuation())
	}
	if page.GetLiveJoinSeq() != 0 {
		t.Fatalf("live_join_seq = %d, want 0", page.GetLiveJoinSeq())
	}
}

func TestPagingStopsAtTheNewestClearOrCompaction(t *testing.T) {
	// Arrange — a clear at seq 10 cut everything below it. History above a
	// clear is history a frontend discards, so paging into it would serve
	// pages nobody can see.
	h := newPageHarness(t, pageTextEvents(t, 12))
	h.floors.SetNewestClearOrCompactSeq("s1", 10)

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 10})

	// Assert — only the items at or above the cut, and the cut IS the
	// beginning.
	got := pageItemUUIDs(page)
	if len(got) != 3 || got[0] != "u10" {
		t.Fatalf("page items = %v, want [u10 u11 u12]", got)
	}
	if page.GetStart() == nil {
		t.Fatalf("page continuation = %v, want the `start` arm: the clear is this conversation's beginning", page.GetContinuation())
	}
}

// --- the clamp --------------------------------------------------------------

func TestAnOversizedLimitIsClampedRatherThanServed(t *testing.T) {
	// Arrange — 80 items, and a client asking for all of them at once. That
	// request IS the full replay this path exists to end.
	h := newPageHarness(t, pageTextEvents(t, 80))

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 5000})

	// Assert.
	if len(page.GetItems()) != pageMaxLimit {
		t.Fatalf("page served %d items for a limit of 5000, want the ceiling of %d", len(page.GetItems()), pageMaxLimit)
	}
}

func TestAZeroLimitTakesTheDaemonDefault(t *testing.T) {
	// Arrange.
	h := newPageHarness(t, pageTextEvents(t, 40))

	// Act.
	page := h.page(t, PageAnchor{Tail: true})

	// Assert.
	if len(page.GetItems()) != pageDefaultLimit {
		t.Fatalf("page served %d items for limit 0, want the default of %d", len(page.GetItems()), pageDefaultLimit)
	}
}

// --- the fence --------------------------------------------------------------

func TestAPageIsRefusedWhenTheEchoedFenceNamesASupersededGeneration(t *testing.T) {
	// Arrange — a tab that outlived a daemon bounce still echoes the
	// generation it was reading when it decided to ask.
	h := newPageHarness(t, pageTextEvents(t, 5))

	// Act.
	_, err := h.m.ConversationPage(context.Background(), "ws", ssm.Fence("s1", "g-retired"), PageAnchor{Tail: true, Limit: 3})

	// Assert — refused BEFORE any read, so a stale request can never
	// half-serve someone else's history.
	if !errors.Is(err, errclass.ErrSessionSuperseded) {
		t.Fatalf("ConversationPage error = %v, want ErrSessionSuperseded", err)
	}
	if replays := h.history.replays(); len(replays) != 0 {
		t.Fatalf("a refused page read %d range(s) from the store, want none", len(replays))
	}
}

func TestAPageStampsTheFenceItWasMintedUnder(t *testing.T) {
	// Arrange — the fence is what a client byte-compares before adopting the
	// page, so an unstamped page is one it must discard.
	h := newPageHarness(t, pageTextEvents(t, 5))

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert — the workspace's PUBLISHED fence, byte for byte.
	if got, want := page.GetFence(), ssm.Fence("s1", "g1"); got != want {
		t.Fatalf("page fence = %q, want the workspace's published fence %q", got, want)
	}
}

func TestAnUngeneratedWorkspacesPageCarriesTheAbsentFenceItPublishes(t *testing.T) {
	// Arrange — the hibernated workspace a cold webview mounts against after a
	// daemon bounce. Its published fence is ABSENT, and a page that composed
	// `Fence(session, "")` instead would stamp a token the mint never produces:
	// the client byte-compares, disagrees, and discards the page it just asked
	// for, which is the blank feed one layer further out.
	h := newPageHarness(t, pageTextEvents(t, 3))
	h.applier.setCurrent("ws", &frontendv1.WorkspaceState{
		Workspace: "ws", SessionId: "s1", ControllerGenerationId: "", Fence: "",
	})

	// Act.
	page, err := h.m.ConversationPage(context.Background(), "ws", "", PageAnchor{Tail: true, Limit: 3})
	if err != nil {
		t.Fatalf("ConversationPage: %v", err)
	}

	// Assert.
	if got := page.GetFence(); got != "" {
		t.Fatalf("page fence = %q, want the absent fence the workspace publishes", got)
	}
}

func TestACursorMintedForAnotherConversationIsRefused(t *testing.T) {
	// Arrange — a cursor whose seq is expressed in a different session's seq
	// space. Serving it would page someone else's history at plausible-looking
	// coordinates.
	h := newPageHarness(t, pageTextEvents(t, 20))
	foreign := encodePageCursor(pageCursor{sessionID: "s-other", beforeSeq: 10})

	// Act.
	_, err := h.m.ConversationPage(context.Background(), "ws", ssm.Fence("s1", "g1"), PageAnchor{Cursor: foreign, Limit: 3})

	// Assert.
	if !errors.Is(err, ErrPageCursorUnreadable) {
		t.Fatalf("ConversationPage error = %v, want ErrPageCursorUnreadable", err)
	}
}

// --- the walk ---------------------------------------------------------------

func TestATailPageDoesNotScanTheWholeConversation(t *testing.T) {
	// Arrange — the whole point. A conversation far larger than one window,
	// asked for its last three items.
	h := newPageHarness(t, pageTextEvents(t, 4000))

	// Act.
	h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert — ONE range read, and it STARTS near the tail rather than at the
	// beginning. The upper bound is deliberately open (a tail page is never
	// capped at the daemon's consumption mark), so what bounds the cost is
	// where the read begins: within one window of the conversation's end.
	replays := h.history.replays()
	if len(replays) != 1 {
		t.Fatalf("tail page issued %d range read(s), want exactly 1", len(replays))
	}
	if replays[0][1] != 0 {
		t.Fatalf("tail page read to_seq = %d, want 0: a tail page's upper bound is open", replays[0][1])
	}
	if from := replays[0][0]; from < 4000-pageInitialWindow {
		t.Fatalf("tail page read from_seq = %d, want it within one window (%d) of the conversation's end at 4000", from, pageInitialWindow)
	}
}

func TestTheWalkWidensUntilThePageIsFull(t *testing.T) {
	// Arrange — one item, sitting far below the tail, with a long stretch of
	// nothing renderable above it. The first window finds nothing, so the walk
	// has to widen rather than serve an empty page.
	events := []*corev1.Event{pageAssistantEvent(t, 1, "u1", []*datav1.ContentBlock{pageTextBlock("m1")})}
	h := newPageHarness(t, events)
	h.seq.SetLastSeq("s1", 3000)

	// Act.
	page := h.page(t, PageAnchor{Tail: true, Limit: 3})

	// Assert — the item is found, and it took more than one window to find it.
	if got := pageItemUUIDs(page); len(got) != 1 || got[0] != "u1" {
		t.Fatalf("page items = %v, want [u1]", got)
	}
	if replays := h.history.replays(); len(replays) < 2 {
		t.Fatalf("the walk issued %d range read(s), want it to have widened at least once", len(replays))
	}
}

File: modules/app/agent-repl/webapp/test/conversation-page.test.ts
import { describe, expect, it } from "vitest";

import { ConversationStore, type ConversationItem, type TextItem } from "../src/store.js";
import type { AdapterEffect, WorkspaceStatusInput } from "../src/state-adapter.js";

/**
 * THE COLD OPEN IS A PAGE, AND THE SPLICE ONTO THE LIVE STREAM IS GAP-FREE.
 *
 * A cold mount used to ask for `from_seq: 0` and receive every store event the
 * session ever produced (259k events / 186MB in the worst workspace observed)
 * to draw a screen whose visible tail is about ten items. These cases pin the
 * replacement: a bounded tail page renders immediately, a load-more walks
 * backwards from it, and the mark the live stream is rejoined at cannot lose
 * an item produced while the page was in flight.
 */

// --- fixtures ---------------------------------------------------------------

function workspaceEffect(over: Partial<WorkspaceStatusInput> = {}): AdapterEffect {
  return {
    kind: "workspace-state",
    value: {
      workspace: "ws",
      sessionId: "s1",
      fence: "f1",
      state: "idle",
      turnActive: false,
      liveTaskCount: 0,
      causeKind: "turn_ended",
      causeSeq: 1,
      atMs: 1000,
      connectivity: "operational",
      sessionStatus: "ready",
      controllerGenerationId: "g1",
      activeFaults: [],
      mergeLeaseHeld: false,
      mergeStatus: null,
      mergeDequeueOffer: null,
      ...over,
    },
  };
}

function textItem(blockId: string, text: string): TextItem {
  return { kind: "text", blockId, messageId: `m-${blockId}`, text, done: true, ts: "2026-01-01T00:00:00Z" };
}

function pageEffect(over: {
  requestId: string;
  items: ConversationItem[];
  continuation?: { case: "more"; cursor: string } | { case: "start" };
  liveJoinSeq?: number;
  fence?: string;
}): AdapterEffect {
  return {
    kind: "conversation-page",
    workspace: "ws",
    fence: over.fence ?? "f1",
    requestId: over.requestId,
    items: over.items,
    continuation: over.continuation ?? { case: "start" },
    liveJoinSeq: over.liveJoinSeq ?? 0,
  };
}

function deltaEffect(items: ConversationItem[], throughSeq: number): AdapterEffect {
  return { kind: "conversation-items", workspace: "ws", fence: "f1", throughSeq, items };
}

/** A store with a live workspace fence, which is what a page is measured against. */
function armedStore(): ConversationStore {
  const store = new ConversationStore();
  store.ingest([workspaceEffect()]);
  return store;
}

const textOf = (store: ConversationStore): string[] =>
  store.state.items.filter((i) => i.kind === "text").map((i) => (i as { text: string }).text);

// --- the tail page ----------------------------------------------------------

describe("the cold-open tail page", () => {
  it("renders its items immediately, oldest first", () => {
    // Arrange — the cold open asked for the newest three.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    // Act
    store.ingest([
      pageEffect({
        requestId: "r-1",
        items: [textItem("a", "18"), textItem("b", "19"), textItem("c", "20")],
        continuation: { case: "more", cursor: "cp1-A" },
        liveJoinSeq: 20,
      }),
    ]);
    // Assert
    expect(textOf(store)).toEqual(["18", "19", "20"]);
  });

  it("adopts live_join_seq as the mark the live stream is rejoined at", () => {
    // Arrange
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    // Act
    store.ingest([pageEffect({ requestId: "r-1", items: [textItem("c", "20")], liveJoinSeq: 20 })]);
    // Assert — the next resync asks from here, INCLUSIVELY.
    expect(store.state.lastSeq).toBe(20);
  });

  it("does not lose an item produced between the page's mint and the subscribe", () => {
    // Arrange — THE GAP-FREE SPLICE. The page is current through seq 20; the
    // session produced seq 21 while the page was still travelling. Because the
    // client rejoins at 20 and the daemon's replay is inclusive of the mark,
    // that item arrives on the very first delta rather than being skipped.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([pageEffect({ requestId: "r-1", items: [textItem("c", "20")], liveJoinSeq: 20 })]);
    // Act — the replay from lastSeq re-delivers 20 (deduped by id) and 21.
    store.ingest([deltaEffect([textItem("c", "20")], 20), deltaEffect([textItem("d", "21")], 21)]);
    // Assert
    expect(textOf(store)).toEqual(["20", "21"]);
    expect(store.state.lastSeq).toBe(21);
  });

  it("an empty conversation pages to nothing without breaking the mark", () => {
    // Arrange — a fresh workspace. Silence is the empty conversation it is.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    // Act
    store.ingest([pageEffect({ requestId: "r-1", items: [], continuation: { case: "start" } })]);
    // Assert
    expect(store.state.items).toHaveLength(0);
    expect(store.state.lastSeq).toBe(0);
    expect(store.state.paging.reachedStart).toBe(true);
  });
});

// --- load-more --------------------------------------------------------------

describe("load-more", () => {
  it("prepends an older page ABOVE everything the feed already holds", () => {
    // Arrange — the tail page is on screen.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([
      pageEffect({
        requestId: "r-1",
        items: [textItem("c", "20")],
        continuation: { case: "more", cursor: "cp1-A" },
        liveJoinSeq: 20,
      }),
    ]);
    // Act
    store.notePageRequested({ requestId: "r-2", anchor: "before", fence: "f1" });
    store.ingest([
      pageEffect({
        requestId: "r-2",
        items: [textItem("a", "18"), textItem("b", "19")],
        continuation: { case: "more", cursor: "cp1-B" },
      }),
    ]);
    // Assert — history reads in conversation order, not arrival order.
    expect(textOf(store)).toEqual(["18", "19", "20"]);
  });

  it("keeps walking down across successive pages", () => {
    // Arrange
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([
      pageEffect({ requestId: "r-1", items: [textItem("c", "3")], continuation: { case: "more", cursor: "A" }, liveJoinSeq: 3 }),
    ]);
    store.notePageRequested({ requestId: "r-2", anchor: "before", fence: "f1" });
    store.ingest([
      pageEffect({ requestId: "r-2", items: [textItem("b", "2")], continuation: { case: "more", cursor: "B" } }),
    ]);
    // Act
    store.notePageRequested({ requestId: "r-3", anchor: "before", fence: "f1" });
    store.ingest([pageEffect({ requestId: "r-3", items: [textItem("a", "1")], continuation: { case: "start" } })]);
    // Assert
    expect(textOf(store)).toEqual(["1", "2", "3"]);
  });

  it("a before page never moves the live mark backwards", () => {
    // Arrange — a load-more is history; adopting a mark from it would make the
    // next resync re-replay everything above it.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([pageEffect({ requestId: "r-1", items: [textItem("c", "20")], liveJoinSeq: 20, continuation: { case: "more", cursor: "A" } })]);
    // Act
    store.notePageRequested({ requestId: "r-2", anchor: "before", fence: "f1" });
    store.ingest([pageEffect({ requestId: "r-2", items: [textItem("b", "19")], continuation: { case: "start" } })]);
    // Assert
    expect(store.state.lastSeq).toBe(20);
  });

  it("the continuation cursor is retained verbatim for the next request", () => {
    // Arrange — opaque: this end stores and returns it, never parses it.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    // Act
    store.ingest([
      pageEffect({ requestId: "r-1", items: [], continuation: { case: "more", cursor: "cp1-OPAQUE" } }),
    ]);
    // Assert
    expect(store.state.paging.cursor).toBe("cp1-OPAQUE");
  });

  it("the conversation's beginning RETIRES the cursor", () => {
    // Arrange
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([pageEffect({ requestId: "r-1", items: [], continuation: { case: "more", cursor: "A" } })]);
    // Act
    store.notePageRequested({ requestId: "r-2", anchor: "before", fence: "f1" });
    store.ingest([pageEffect({ requestId: "r-2", items: [], continuation: { case: "start" } })]);
    // Assert
    expect(store.state.paging.reachedStart).toBe(true);
    expect(store.state.paging.cursor).toBeNull();
  });
});

// --- refusals ---------------------------------------------------------------

describe("a page the store will not adopt", () => {
  it("a stale fence discards the page WHOLE and marks it for re-request", () => {
    // Arrange — a page minted under a generation this client no longer reads.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    // Act
    store.ingest([
      pageEffect({
        requestId: "r-1",
        items: [textItem("a", "18"), textItem("b", "19")],
        fence: "f-retired",
        liveJoinSeq: 20,
      }),
    ]);
    // Assert — not one item of it was adopted, and the mark did not move.
    expect(store.state.items).toHaveLength(0);
    expect(store.state.lastSeq).toBe(0);
    expect(store.state.paging.staleFenceRequestId).toBe("r-1");
  });

  it("a page answering a request this client abandoned is discarded", () => {
    // Arrange — adopting it would splice history in at a cursor the feed no
    // longer holds.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.forgetPageRequest("r-1");
    // Act
    store.ingest([pageEffect({ requestId: "r-1", items: [textItem("a", "18")], liveJoinSeq: 20 })]);
    // Assert
    expect(store.state.items).toHaveLength(0);
  });

  it("a page whose id is not the outstanding one is discarded", () => {
    // Arrange — a cold open and a load-more can both be in flight; only the
    // echo distinguishes their answers.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-2", anchor: "before", fence: "f1" });
    // Act
    store.ingest([pageEffect({ requestId: "r-1", items: [textItem("a", "18")] })]);
    // Assert
    expect(store.state.items).toHaveLength(0);
  });
});

// --- the re-anchor ----------------------------------------------------------

/**
 * THE RECONNECT WITH A RETIRED MARK: the tail page REPLACES the conversation.
 *
 * A page that survived a vendor session uuid rotation holds items ranked in a
 * store seq space that no longer exists and a mark counted in it. The daemon
 * refuses that mark (`rejection_cause=retired_seq_space`) instead of flooring it
 * into a whole-conversation replay, and this end re-anchors: `rebaseSeqSpace`
 * drops the retired conversation, and a tail page supplies a new one.
 *
 * APPENDING WOULD BE THE QUIET WRONG ANSWER. The new space starts again at 1, so
 * its items would rank BELOW a thousand items they actually follow, and the
 * retired mark — being the larger number — would survive the `>` guard on
 * `lastSeq` and keep the next resync asking from the dead space forever.
 */
describe("re-anchoring after a retired replay mark", () => {
  it("replaces the retired conversation rather than appending to it", () => {
    // Arrange — a feed full of the retired space's history.
    const store = armedStore();
    store.ingest([deltaEffect([textItem("b-old-1", "old one"), textItem("b-old-2", "old two")], 1060)]);
    store.rebaseSeqSpace();
    // Act — the tail page of the LIVE space.
    store.notePageRequested({ requestId: "r-anchor", anchor: "tail", fence: "f1" });
    store.ingest([
      pageEffect({ requestId: "r-anchor", items: [textItem("b-new", "new one")], liveJoinSeq: 12 }),
    ]);
    // Assert
    expect(textOf(store)).toEqual(["new one"]);
  });

  it("adopts the live space's join seq even though it is BELOW the retired mark", () => {
    // Arrange — 12 is smaller than 1060, which is exactly why the retired mark
    // has to be dropped rather than compared against.
    const store = armedStore();
    store.ingest([deltaEffect([textItem("b-old-1", "old one")], 1060)]);
    store.rebaseSeqSpace();
    // Act
    store.notePageRequested({ requestId: "r-anchor", anchor: "tail", fence: "f1" });
    store.ingest([
      pageEffect({ requestId: "r-anchor", items: [textItem("b-new", "new one")], liveJoinSeq: 12 }),
    ]);
    // Assert
    expect(store.state.lastSeq).toBe(12);
  });

  it("drops the retired conversation's load-more cursor", () => {
    // Arrange — the cursor is the daemon's handle on a position in a
    // conversation that is gone.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([
      pageEffect({
        requestId: "r-1",
        items: [textItem("b-old", "old one")],
        continuation: { case: "more", cursor: "c-retired" },
        liveJoinSeq: 1060,
      }),
    ]);
    // Act
    store.rebaseSeqSpace();
    // Assert
    expect(store.state.paging.cursor).toBeNull();
  });

  it("un-retires the load-more affordance the retired conversation had spent", () => {
    // Arrange — `reachedStart` is a fact established about the OLD conversation.
    const store = armedStore();
    store.notePageRequested({ requestId: "r-1", anchor: "tail", fence: "f1" });
    store.ingest([
      pageEffect({ requestId: "r-1", items: [textItem("b-old", "old one")], liveJoinSeq: 1060 }),
    ]);
    // Act
    store.rebaseSeqSpace();
    // Assert
    expect(store.state.paging.reachedStart).toBe(false);
  });
});
